src/Microsoft.ApiDesignGuidelines.Analyzers/Core/IdentifiersShouldNotContainUnderscores.cs

```diff
@@ -31,78 +35,266 @@ public abstract class IdentifiersShouldNotContainUnderscoresAnalyzer : Diagnosti
+                                {
+                                    symbolAnalysisContext.ReportDiagnostic(symbol.CreateDiagnostic(NamespaceRule, symbol.ToDisplayString()));
+                                    return;
+                                }
+
+                            case SymbolKind.NamedType:
+                                {
+                                    if (!symbol.IsPublic())
+                                    {
+                                        return;
+                                    }
+
+                                    symbolAnalysisContext.ReportDiagnostic(symbol.CreateDiagnostic(TypeRule, symbol.ToDisplayString()));
+                                    return;
+                                }
+
+                            case SymbolKind.Field:
+                                {
+                                    var fieldSymbol = symbol as IFieldSymbol;
+                                    if (symbol.IsPublic() && (fieldSymbol.IsConst || (fieldSymbol.IsStatic && fieldSymbol.IsReadOnly)))
+                                    {
+                                        symbolAnalysisContext.ReportDiagnostic(symbol.CreateDiagnostic(MemberRule, symbol.ToDisplayString(SymbolDisplayFormat.CSharpErrorMessageFormat)));
+                                        return;
+                                    }
+
+                                    return;
+                                }
+
+                            default:
+                                {
+                                    if (IsInvalidSymbol(symbol))
+                                    {
+                                        return;
+                                    }
+
+                                    symbolAnalysisContext.ReportDiagnostic(symbol.CreateDiagnostic(MemberRule, symbol.ToDisplayString(SymbolDisplayFormat.CSharpErrorMessageFormat)));
+                                    return;
+                                }
+                        }
+                    }
+                },
+                SymbolKind.Namespace, // Namespace
+                SymbolKind.NamedType, //Type
+                SymbolKind.Method, SymbolKind.Property, SymbolKind.Field, SymbolKind.Event // Members
+                );
+
+                GetSyntaxNodeDiagnostics(compilationStartAnalysisContext);
+            });
+
+            analysisContext.RegisterCompilationAction(compilationAnalysisContext =>
+            {
+                var compilation = compilationAnalysisContext.Compilation;
+                if (ContainsUnderScore(compilation.AssemblyName))
+                {
+                    compilationAnalysisContext.ReportDiagnostic(compilation.Assembly.CreateDiagnostic(AssemblyRule, compilation.AssemblyName));
+                }
+            });
+        }
+
+        private bool IsInvalidSymbol(ISymbol symbol)
+        {
+            return (!((symbol.IsPublic() || symbol.IsProtected()) && !symbol.IsOverride)) ||
```

Comment:
This seems like checking for effective visibility. Can't you use `ISymbolExtensions.GetResultantVisibility`?